web/src/search_suggestion.ts

```diff
@@ -729,46 +729,78 @@ function get_operator_suggestions(last: NarrowTerm): Suggestion[] {
     });
 }
 
+// One full search suggestion can include multiple search terms, based on what's
+// in the search bar.
+type SuggestionLine = Suggestion[];
+function suggestion_search_string(suggestion_line: SuggestionLine): string {
+    const search_strings = [];
+    for (const suggestion of suggestion_line) {
+        if (suggestion.search_string !== "") {
+            search_strings.push(suggestion.search_string);
+        }
+    }
+    return search_strings.join(" ");
+}
+
 class Attacher {
-    result: Suggestion[] = [];
+    result: SuggestionLine[] = [];
     prev = new Set<string>();
-    base: Suggestion;
+    base: SuggestionLine;
 
-    constructor(base: Suggestion) {
+    constructor(base: SuggestionLine) {
         this.base = base;
     }
 
-    prepend_base(suggestion: Suggestion): void {
-        if (this.base && this.base.description_html.length > 0) {
-            suggestion.search_string = this.base.search_string + " " + suggestion.search_string;
-            suggestion.description_html =
-                this.base.description_html + ", " + suggestion.description_html;
-        }
-    }
-
-    push(suggestion: Suggestion): void {
-        if (!this.prev.has(suggestion.search_string)) {
-            this.prev.add(suggestion.search_string);
-            this.result.push(suggestion);
+    push(suggestion_line: SuggestionLine): void {
+        const search_string = suggestion_search_string(suggestion_line);
+        if (!this.prev.has(search_string)) {
+            this.prev.add(search_string);
+            this.result.push(suggestion_line);
         }
     }
 
     push_many(suggestions: Suggestion[]): void {
         for (const suggestion of suggestions) {
-            this.push(suggestion);
+            this.push([suggestion]);
         }
     }
 
     attach_many(suggestions: Suggestion[]): void {
         for (const suggestion of suggestions) {
-            this.prepend_base(suggestion);
-            this.push(suggestion);
+            const suggestion_line = [...this.base, suggestion];
+            this.push(suggestion_line);
         }
     }
+
+    get_result(): Suggestion[] {
+        return this.result.map((suggestion_line) => {
+            const description_htmls = [];
+            const search_strings = [];
+            for (const suggestion of suggestion_line) {
+                if (suggestion.is_person) {
+                    assert(suggestion.description_html !== "");
+                    description_htmls.push(
+                        suggestion.description_html +
+                            render_user_pill({user_pill_context: suggestion.user_pill_context!}),
+                    );
```

Comment:
Does this change to use `render_user_pill` have any user-facing impact? It's kinda hard to follow, but it does look to be equivalent:

```
$ cat web/templates/search_list_item.hbs
<div class="search_list_item">
    <span>{{{ description_html }}}</span>
    {{#if is_person}}
    <span class="pill-container pill-container-btn">
        {{> input_pill user_pill_context}}
    </span>
    {{/if}}
</div>
```

Reply:
This would be the exact equivalency of what we were doing before (replacing what happened in `prepend_base`)

```js
    get_result(): Suggestion[] {
        return this.result.map((suggestion_line) => {
            const description_htmls = [];
            const search_strings = [];
            for (const suggestion of suggestion_line) {
                if (suggestion.description_html !== "") {
                    description_htmls.push(suggestion.description_html);
                }

                if (suggestion.search_string !== "") {
                    search_strings.push(suggestion.search_string);
                }
            }
            const last_suggestion = suggestion_line.at(-1);
            if (last_suggestion?.is_person) {
                const user_pill_context = last_suggestion.user_pill_context
                assert(user_pill_context !== undefined);
                return {
                    description_html: description_htmls.join(", "),
                    search_string: search_strings.join(" "),
                    is_person: true,
                    user_pill_context,
                };
            }
            return {
                description_html: description_htmls.join(", "),
                search_string: search_strings.join(" "),
            };
        });
    }
```

Maybe it makes more sense to do it this way instead for this PR -- but at some point we're going to need to get a way to render suggestion lines with multiple user pills in them (#30426). I can work on that in #30426 instead, and maybe that should involve changing the template code instead of code here. 

Thanks for pointing it out! I haven't noticed any duplicate `display_html`, but I agree it's confusing regardless so I'll look into it more.

Reply:
Okay yeah it's clearer to see what changes are needed if you have the context of the rest of this commit.

https://github.com/zulip/zulip/pull/30426/commits/0193932d5a26065e43118afe3feb731bf475139c#diff-426bbd5e4a5470f26a21e1b64d96f3d668b91dc2602cb51c77786126162374f2R833

("search_suggestion: Show pills for all users in a group suggestion." in #30426)